### PR TITLE
Use ajax power on module order field

### PR DESCRIPTION
**Pull request by dgrammatiko**
#### Enhance the current status quo

Try to create a new module in module manager. Select a position. Select another position
The dropdown ordering doesn't really represent the true order of the modules in the new position.
So my idea was to use ajax power to be always accurate on that dropdown.
#### Summary of Changes

Use ajax power, tight connection with the controller module.php
#### Preview

Mind the untranslated string!
No other module in that position:
screen shot 2016-07-06 at 23 23 04

With other modules in the same position:
screen shot 2016-07-06 at 23 24 48
#### Testing

Please DO NOT TEST this one!
Instead give a vote (up or down thumb) so PLT can make a decision (or PLT decides directly, whatever comes first)

### CODE REVIEW

File: libraries/cms/form/field/moduleorder.php [hunk 1]
```diff
@@ -44,55 +43,51 @@ protected function getInput()
...
+			var getNewOrder = function() {
+				$.ajax({
+					type: "GET",
+					dataType: "json",
+					url: "index.php?option=com_modules&task=module.orderPosition&$token",
+					data: { "client_id": client_id,
+							"position": originalPos
+					 },
+					success:function(response) {
+					 	console.log(response.data);
```
Comment (Devportobello) [inline libraries/cms/form/field/moduleorder.php]: debug code ?

File: media/system/js/core.js [hunk 2]
```diff
@@ -1 +1,636 @@
...mitform=function(t){e.submitform(t)},window.saveorder=function(e,t){window.checkAll_button(e,t)},window.checkAll_button=function(n,o){o=o?o:"saveorder";var r,i;for(r=0;n>=r;r++){if(i=t.adminForm["cb"+r],!i)return void alert("You cannot change the order of items, as an item in the list is `Checked Out`");i.checked=!0}e.submitform(o)},e.loadingLayer=function(n,o){if(n=n||"show",o=o||t.body,"load"==n){var r=t.getElementsByTagName("body")[0].getAttribute("data-basepath")||"",i=t.createElement("div");i.id="loading-logo",i.style.position="fixed",i.style.top="0",i.style.left="0",i.style.width="100%",i.style.height="100%",i.style.opacity="0.8",i.style.filter="alpha(opacity=80)",i.style.overflow="hidden",i.style["z-index"]="10000",i.style.display="none",i.style["background-color"]="#fff",i.style["background-image"]='url("'+r+'/media/jui/images/ajax-loader.gif")',i.style["background-position"]="center",i.style["background-repeat"]="no-repeat",i.style["background-attachment"]="fixed",o.appendChild(i)}else t.getElementById("loading-logo")||e.loadingLayer("load",o),t.getElementById("loading-logo").style.display="show"==n?"block":"none";return t.getElementById("loading-logo")}}(Joomla,document);
```
Comment (Devportobello) [inline media/system/js/core.js]: The following code are not minified, so this is for core-uncompressed.js :)

File: libraries/cms/form/field/moduleorder.php (hunk 1 above)
Comment (dgt41) [inline libraries/cms/form/field/moduleorder.php]: This PR is just to showcase some working code. it's not commit ready. 
Eg `$("#jform_position")` the relation with the position field right now is hardcoded, needs an option in the XML for the field moduleorder

File: administrator/components/com_modules/controllers/module.php [hunk 3]
```diff
@@ -224,4 +224,80 @@ public function save($key = null, $urlVar = null)
...
+			for ($i = 0, $n; $i < $n; $i++)
+			{
+				if (!isset($orders2[$orders[$i]->position]))
+				{
+					$orders2[$orders[$i]->position] = 0;
+				}
+
+				$orders2[$orders[$i]->position]++;
+				$ord = $orders2[$orders[$i]->position];
+				$title = JText::sprintf('COM_MODULES_OPTION_ORDER_POSITION', $ord, addslashes($orders[$i]->title));
```
Comment (Devportobello) [inline administrator/components/com_modules/controllers/module.php]: RFC i know, just to advice: addslashes give me: "11. Who\'s Online"

**Comment by Devportobello** [comment]
Dunno if people use the ordering directly in edit mode, but this is an improvment when there is many module in same position

File: administrator/components/com_modules/controllers/module.php (hunk 3 above)
Comment (dgt41) [inline administrator/components/com_modules/controllers/module.php]: That was a blind transfer of the old code, nice finding I will fix that!

**Comment by dgt41** [comment]
@Devportobello from a UX perspective this is the right approach.
Now is this something that people actually use? I have no clue...

**Comment by RonakParmar** [comment]
I have tested this item :red_circle: unsuccessfully on 4885e07250ca96c2dfce70705b733b1119a22ec3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Comment by RonakParmar** [comment]
I have applied this PR and try to create "custom" module, I have jQuery error in console.
Joomla Version: Joomla! 3.6.0-rc Release Candidate [ Noether ] 28-June-2016 20:34 GMT 
Tested in Ubuntu 12.04, FireFox 47.0.1.screen shot 2016-07-12 at 07 15 42<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Comment by killoltailored** [comment]
I have tested this item :red_circle: unsuccessfully on 4885e07250ca96c2dfce70705b733b1119a22ec3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Comment by killoltailored** [comment]
I have tested after applying patch with Joomla Version Joomla! 3.6.0-rc Release Candidate [ Noether ] 28-June-2016 20:34 GMT in Windows 8.1, Firefox 47.0.1
Create new custom module and found jQuery error. Test with both case select unused position and used position.screen shot 2016-07-12 at 07 27 03screen shot 2016-07-12 at 07 27 06<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Comment by dgt41** [comment]
@RonakParmar @killoltailored Thanks for testing but this is RFC (Request for comments) which means a decision is needed here before I spent more time on this pr. So an up vote or down vote on the scope is enough at the moment (even if the code is not show time ready). Thanks again

**Comment by dgt41** [comment]
@brianteeman I thought that RFC indicator was enough. Maybe I should have opened this as an issue without code attached so people wouldn't be able to do any testing.

**Comment by brianteeman** [comment]
Don't forget that we have a lot of less experienced developers and new
testers.

Never assume anything - when you assume you make a s ASS out of U and ME

**Comment by gunjanpatel** [comment]
I am sure they are testing it because they see `pending` status of the task. It is hard to understand what is RFC from the title of the task. :smile:  

It isn't should be in a "Needs Review" status? Or may just "New" status and which waiting to be "Confirmed". 

Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Comment by brianteeman** [comment]
Gunman. Unfortunately any PR will show up in patch tester. Patchtester only
shows if it is already RTC or not. Needs Review is an internal flag for CMS
maintainers to make a decision on something.

To be honest a comment that it didn't work is a reply to a request for a
comment in my book :)

**Comment by JoshuaLewis** [comment]
Great idea. So who are you asking if this is worth it before working on this PR? The folks who manage PR's for new releases of Joomla? Testers?

**Comment by brianteeman** [comment]
@dgt41 - seems the consensus is that yes this is a good idea but that the code currently doesnt work yet<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Comment by dgt41** [comment]
Now it should be testable

**Comment by gunjanpatel** [comment]
Can you remove RFC from title then?

**Comment by brianteeman** [comment]
Not sure if it is doing what I expect it to do
After applying the patch the ordering in the select is correct but when viewing it on the front end the ordering selected is not being used<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Comment by dgt41** [comment]
@brianteeman before applying this patch was the behaviour different? All this pr is doing is reloading the ordering when the position is changed, it shouldn't interfere with the actual ordering in the front end...

File: libraries/cms/form/field/moduleorder.php [hunk 4]
```diff
@@ -19,80 +19,123 @@ class JFormFieldModuleOrder extends JFormField
...
+	/**
+	 * Method to get the field input markup for the file field.
+	 * Field attributes allow specification of a maximum file size and a string
+	 * of accepted file extensions.
+	 *
+	 * @return  string  The field input markup.
+	 *
+	 * @note    The field does not include an upload mechanism.
+	 * @see     JFormFieldMedia
+	 * @since   11.1
```
Comment (andrepereiradasilva) [inline libraries/cms/form/field/moduleorder.php]: `11.1`? better always to use `__DEPLOY_VERSION__`

Comment (gunjanpatel) [inline libraries/cms/form/field/moduleorder.php]: I think `__DEPLOY_VERSION__` needs to be on other places too where `3.7` is added.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 9ed547eb37f514d12c3c03859a1f4d18d75a6f40<br /><br />No idea what the issue was I saw before as I cant replicate it now

I still dont really see what the difference is now though compared to before this pr<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

File: libraries/cms/form/field/moduleorder.php (hunk 4 above)
Comment (dgt41) [inline libraries/cms/form/field/moduleorder.php]: @andrepereiradasilva @gunjanpatel you are both right but I think `__DEPLOY_VERSION__` came after july 6th. Anyways I will fix this..

Comment (gunjanpatel) [inline libraries/cms/form/field/moduleorder.php]: ohh sorry didn't noticed dates. Thanks :+1:

**Comment by AnishaTailored** [comment]
I have tested this item :white_check_mark: successfully on 9ed547eb37f514d12c3c03859a1f4d18d75a6f40<br /><br />I have tested this successfully. If we change the position then the ordering field fetches modules of the newly selected position.
But shouldn't Ordering dropdown be jQuery chosen dropdown as it was before this PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Comment by zero-24** [comment]
@dgt41 can you double check the since tags? If that is done i think we can RTC

**Comment by dgt41** [comment]
@zero-24 there you go!

File: libraries/cms/form/field/moduleorder.php [hunk 5]
```diff
@@ -19,80 +19,119 @@ class JFormFieldModuleOrder extends JFormField
...
+	}
 
-			$html[] = 'orders[' . $i . '] =  new Array("' . $orders[$i]->position . '","' . $ord . '","' . $title . '");';
-		}
+	/**
+	 * Method to get the field input markup for the moduleorder field.
+	 *
+	 * @return  string  The field input markup.
+	 *
+	 * @since   11.1
```
Comment (zero-24) [inline libraries/cms/form/field/moduleorder.php]: I think this is missing?

File: libraries/cms/form/field/moduleorder.php [hunk 6]
```diff
@@ -19,80 +19,119 @@ class JFormFieldModuleOrder extends JFormField
...
+	 * @param   SimpleXMLElement  $element  The SimpleXMLElement object representing the `<field>` tag for the form field object.
+	 * @param   mixed             $value    The form field value to validate.
+	 * @param   string            $group    The field name group control value. This acts as as an array container for the field.
+	 *                                      For example if the field has name="foo" and the group value is set to "bar" then the
+	 *                                      full field name would end up being "bar[foo]".
+	 *
+	 * @return  boolean  True on success.
+	 *
+	 * @see     JFormField::setup()
+	 * @since   3.2
```
Comment (zero-24) [inline libraries/cms/form/field/moduleorder.php]: And this one?

File: administrator/components/com_modules/controllers/module.php [hunk 7]
```diff
@@ -224,4 +224,80 @@ public function save($key = null, $urlVar = null)
...
+				$title = JText::sprintf('COM_MODULES_OPTION_ORDER_POSITION', $ord, htmlspecialchars($orders[$i]->title, ENT_QUOTES, 'UTF-8'));
+
+				$html[] = $orders[$i]->position . ',' . $ord . ',' . $title;
+			}
+		}
+		else
+		{
+			$html[] = $position . ',' . 1 . ',' . JText::_('JNONE');
+		}
+
```
Comment (zero-24) [inline administrator/components/com_modules/controllers/module.php]: Here is just one space to much ;)

**Comment by zero-24** [comment]
Hmm i have found two more ;)

File: libraries/cms/form/field/moduleorder.php (hunk 5 above)
Comment (dgt41) [inline libraries/cms/form/field/moduleorder.php]: getInput() was there before so the since should not change. (11.1 is wrong should be 1.6)

**Comment by zero-24** [comment]
Go for it @joomla-cms-bot please make us happy :smile:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

**Pull request merged**